// @supabase/postgres-meta/dist/server/utils.js
import pgcs from 'pg-connection-string';
import { DEFAULT_POOL_CONFIG } from './constants.js';
export const extractRequestForLogging = (request) => {
    let pg = 'unknown';
    try {
        if (request.headers.pg) {
            pg = pgcs.parse(request.headers.pg).host || pg;
        }
    }
    catch (e) {
        console.warn('failed to parse PG connstring for ' + request.url);
    }
    const additional = request.headers['x-supabase-info']?.toString() || '';
    return {
        method: request.method,
        url: request.url,
        pg,
        opt: additional,
    };
};
export function createConnectionConfig(request) {
    const connectionString = request.headers.pg;
    const config = { ...DEFAULT_POOL_CONFIG, connectionString };
    // Override application_name if custom one provided in header
    if (request.headers['x-pg-application-name']) {
        config.application_name = request.headers['x-pg-application-name'];
    }
    return config;
}
export function translateErrorToResponseCode(error, defaultResponseCode = 400) {
    if (error.message === 'Connection terminated due to connection timeout') {
        return 504;
    }
    else if (error.message === 'sorry, too many clients already') {
        return 503;
    }
    else if (error.message === 'Query read timeout') {
        return 408;
    }
    return defaultResponseCode;
}
