// @supabase/postgres-meta/dist/lib/Parser.js
import prettier from 'prettier/standalone.js';
import SqlFormatter from 'prettier-plugin-sql';
import { parse, deparse } from 'pgsql-parser';
const DEFAULT_FORMATTER_OPTIONS = {
    plugins: [SqlFormatter],
    formatter: 'sql-formatter',
    language: 'postgresql',
    database: 'postgresql',
    parser: 'sql',
};
/**
 * Parses a SQL string into an AST.
 */
export function Parse(sql) {
    try {
        const data = parse(sql);
        return { data, error: null };
    }
    catch (error) {
        return { data: null, error: error };
    }
}
/**
 * Deparses an AST into SQL string.
 */
export function Deparse(parsedSql) {
    try {
        const data = deparse(parsedSql, {});
        return { data, error: null };
    }
    catch (error) {
        return { data: null, error: error };
    }
}
/**
 * Formats a SQL string into a prettier-formatted SQL string.
 */
export async function Format(sql, options = {}) {
    try {
        const formatted = await prettier.format(sql, {
            ...DEFAULT_FORMATTER_OPTIONS,
            ...options,
        });
        return { data: formatted, error: null };
    }
    catch (error) {
        return { data: null, error: error };
    }
}
