// @supabase/postgres-meta/dist/lib/PostgresMetaViews.js
import { DEFAULT_SYSTEM_SCHEMAS } from './constants.js';
import { coalesceRowsToArray, filterByList, filterByValue } from './helpers.js';
import { VIEWS_SQL } from './sql/views.sql.js';
import { COLUMNS_SQL } from './sql/columns.sql.js';
export default class PostgresMetaViews {
    query;
    constructor(query) {
        this.query = query;
    }
    async list({ includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset, includeColumns = true, } = {}) {
        const schemaFilter = filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : undefined);
        const sql = generateEnrichedViewsSql({ includeColumns, schemaFilter, limit, offset });
        return await this.query(sql);
    }
    async retrieve({ id, name, schema = 'public', }) {
        if (id) {
            const idsFilter = filterByValue([id]);
            const sql = generateEnrichedViewsSql({
                includeColumns: true,
                idsFilter,
            });
            const { data, error } = await this.query(sql);
            if (error) {
                return { data, error };
            }
            else if (data.length === 0) {
                return { data: null, error: { message: `Cannot find a view with ID ${id}` } };
            }
            else {
                return { data: data[0], error };
            }
        }
        else if (name) {
            const viewIdentifierFilter = filterByValue([`${schema}.${name}`]);
            const sql = generateEnrichedViewsSql({
                includeColumns: true,
                viewIdentifierFilter,
            });
            const { data, error } = await this.query(sql);
            if (error) {
                return { data, error };
            }
            else if (data.length === 0) {
                return {
                    data: null,
                    error: { message: `Cannot find a view named ${name} in schema ${schema}` },
                };
            }
            else {
                return { data: data[0], error };
            }
        }
        else {
            return { data: null, error: { message: 'Invalid parameters on view retrieve' } };
        }
    }
}
const generateEnrichedViewsSql = ({ includeColumns, schemaFilter, idsFilter, viewIdentifierFilter, limit, offset, }) => `
with views as (${VIEWS_SQL({ schemaFilter, limit, offset, viewIdentifierFilter, idsFilter })})
  ${includeColumns ? `, columns as (${COLUMNS_SQL({ schemaFilter, tableIdentifierFilter: viewIdentifierFilter, tableIdFilter: idsFilter })})` : ''}
select
  *
  ${includeColumns ? `, ${coalesceRowsToArray('columns', 'columns.table_id = views.id')}` : ''}
from views`;
