// pgsql-deparser/v13-to-v17-direct.js
"use strict";
/**
 * Auto-generated file with types stripped for better tree-shaking
 * DO NOT EDIT - Generated by strip-direct-transformer-types.ts
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PG13ToPG17Transformer = void 0;
const v13_to_v14_1 = require("./v13-to-v14");
const v14_to_v15_1 = require("./v14-to-v15");
const v15_to_v16_1 = require("./v15-to-v16");
const v16_to_v17_1 = require("./v16-to-v17");
/**
 * Direct transformer from PG13 to PG17
 * This transformer chains v13->v14->v15->v16->v17 transformations
 */
class PG13ToPG17Transformer {
    v13to14 = new v13_to_v14_1.V13ToV14Transformer();
    v14to15 = new v14_to_v15_1.V14ToV15Transformer();
    v15to16 = new v15_to_v16_1.V15ToV16Transformer();
    v16to17 = new v16_to_v17_1.V16ToV17Transformer();
    transform(node) {
        // If it's a ParseResult, handle it specially
        if (this.isParseResult(node)) {
            // Transform through the chain: v13->v14->v15->v16->v17
            const v14Stmts = node.stmts.map((stmtWrapper) => {
                if (stmtWrapper.stmt) {
                    const v14Stmt = this.v13to14.transform(stmtWrapper.stmt, { parentNodeTypes: [] });
                    return { ...stmtWrapper, stmt: v14Stmt };
                }
                return stmtWrapper;
            });
            const v15Stmts = v14Stmts.map((stmtWrapper) => {
                if (stmtWrapper.stmt) {
                    const v15Stmt = this.v14to15.transform(stmtWrapper.stmt, { parentNodeTypes: [] });
                    return { ...stmtWrapper, stmt: v15Stmt };
                }
                return stmtWrapper;
            });
            const v16Stmts = v15Stmts.map((stmtWrapper) => {
                if (stmtWrapper.stmt) {
                    const v16Stmt = this.v15to16.transform(stmtWrapper.stmt, { parentNodeTypes: [] });
                    return { ...stmtWrapper, stmt: v16Stmt };
                }
                return stmtWrapper;
            });
            const v17Stmts = v16Stmts.map((stmtWrapper) => {
                if (stmtWrapper.stmt) {
                    const v17Stmt = this.v16to17.transform(stmtWrapper.stmt, { parentNodeTypes: [] });
                    return { ...stmtWrapper, stmt: v17Stmt };
                }
                return stmtWrapper;
            });
            return {
                ...node,
                version: 170004, // PG17 version
                stmts: v17Stmts
            };
        }
        // Otherwise, transform as a regular node through the chain
        const v14Node = this.v13to14.transform(node, { parentNodeTypes: [] });
        const v15Node = this.v14to15.transform(v14Node, { parentNodeTypes: [] });
        const v16Node = this.v15to16.transform(v15Node, { parentNodeTypes: [] });
        return this.v16to17.transform(v16Node, { parentNodeTypes: [] });
    }
    /**
     * Transform a single statement from PG13 to PG17
     * @deprecated Use transform() instead, which handles all node types
     */
    transformStatement(stmt) {
        const v14Stmt = this.v13to14.transform(stmt, { parentNodeTypes: [] });
        const v15Stmt = this.v14to15.transform(v14Stmt, { parentNodeTypes: [] });
        const v16Stmt = this.v15to16.transform(v15Stmt, { parentNodeTypes: [] });
        return this.v16to17.transform(v16Stmt, { parentNodeTypes: [] });
    }
    /**
     * Type guard to check if a node is a ParseResult
     */
    isParseResult(node) {
        return node && typeof node === 'object' && 'version' in node && 'stmts' in node;
    }
}
exports.PG13ToPG17Transformer = PG13ToPG17Transformer;
